firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsReportDataCapture.java

```diff
@@ -0,0 +1,432 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.crashlytics.internal.common;
+
+import android.app.ActivityManager;
+import android.app.ActivityManager.RunningAppProcessInfo;
+import android.content.Context;
+import android.os.Build;
+import android.os.Build.VERSION;
+import android.os.Environment;
+import android.os.StatFs;
+import com.google.firebase.crashlytics.BuildConfig;
+import com.google.firebase.crashlytics.internal.model.Architecture;
+import com.google.firebase.crashlytics.internal.stacktrace.StackTraceTrimmingStrategy;
+import com.google.firebase.crashlytics.internal.stacktrace.TrimmedThrowableData;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * This class is responsible for capturing information from the system and exception objects,
+ * parsing them, and passing them to handlers implemented by callers to use the captured information
+ * as they see fit.
+ */
+public class CrashlyticsReportDataCapture {
+
+  /**
+   * Defines methods for handling static report data gathered from the device, as well as system and
+   * build properties.
+   */
+  public interface ReportDataHandler {
+    void startReportDataCapture(
+        String sdkVersion,
+        String gmpAppId,
+        String installationId,
+        String buildVersion,
+        String displayVersion);
+
+    void startSessionDataCapture(String generator);
+
+    void startSessionApplicationDataCapture(
+        String packageName,
+        String versionCode,
+        String versionName,
+        String installUuid); // TODO: Unity version
+
+    void endSessionApplicationDataCapture();
+
+    void startSessionOSDataCapture(String osRelease, String osCodeName, boolean isRooted);
+
+    void endSessionOSDataCapture();
+
+    void startSessionDeviceDataCapture(
+        int arch,
+        String model,
+        int availableProcessors,
+        long totalRam,
+        long diskSpace,
+        boolean isEmulator,
+        int state,
+        String manufacturer,
+        String modelClass);
+
+    void endSessionDeviceDataCapture();
+
+    void endSessionDataCapture();
+
+    void endReportDataCapture();
+  }
+
+  /** Defines methods for handling data for a given exception event. */
```

Comment:
It seems like there should be some lifecycle that we should document. I mean, I assume that handlers can assume that a start comes before and end for the same event. But is there more to. Are there other assumptions about the order of events that we should make explicit?

Reply:
New approach will eliminate this concern.